cmd/cstor-pool-mgmt/controller/replica-controller/new_replica_controller.go

```diff
@@ -59,119 +60,201 @@ type CStorVolumeReplicaController struct {
-			} else {
-				q.Operation = common.QOpModify
-				glog.Infof("cStorVolumeReplica Modify event : %v, %v", newCVR.ObjectMeta.Name, string(newCVR.ObjectMeta.UID))
-				controller.recorder.Event(newCVR, corev1.EventTypeNormal, string(common.SuccessSynced), string(common.MessageModifySynced))
-				return // will be removed once modify is implemented
-			}
-			controller.enqueueCStorReplica(newCVR, q)
-		},
-		DeleteFunc: func(obj interface{}) {
-			cVR := obj.(*apis.CStorVolumeReplica)
-			if !IsRightCStorVolumeReplica(cVR) {
-				return
-			}
-			glog.Infof("cVR Resource deleted event: %v, %v", cVR.ObjectMeta.Name, string(cVR.ObjectMeta.UID))
-		},
-	})
+	glog.Info("will set up informer event handlers for cvr")
+
+	ql := common.QueueLoad{}
+
+	cvrInformer.Informer().
+		AddEventHandler(cache.ResourceEventHandlerFuncs{
+			AddFunc: func(obj interface{}) {
+				cvrObj := obj.(*apis.CStorVolumeReplica)
+
+				glog.V(4).Infof(
+					"received informer add event for cvr {%s}",
+					cvrObj.Name,
+				)
+
+				if !IsRightCStorVolumeReplica(cvrObj) || IsErrorDuplicate(cvrObj) {
+					// do nothing
+					return
+				}
+
+				controller.recorder.Event(
+					cvrObj,
+					corev1.EventTypeNormal,
+					string(common.SuccessSynced),
+					string(common.MessageCreateSynced),
+				)
+
+				// new cvr requests will have phase as blank
+				//
+				// NOTE:
+				//  for every restart of controller container
+				// this informer handler will get add event
+				// for each cvr resource present in k8s
+				if IsEmptyStatus(cvrObj) {
+					cvrObj.Status.Phase = apis.CVRStatusInit
+				} else {
+					cvrObj.Status.Phase = apis.CVRStatusRecreate
+				}
+
+				cvrObj, _ = controller.clientset.
+					OpenebsV1alpha1().
+					CStorVolumeReplicas(cvrObj.Namespace).
+					Update(cvrObj)
+
+					// push this operation to workqueue
+				ql.Operation = common.QOpAdd
+				controller.enqueueCStorReplica(cvrObj, ql)
+			},
+
+			// TODO @amitkumardas
+			//
+			// Need to think of writing more manageable code
+			// In the current code, ordering of conditions
+			// is very important. I am sure these conditions
+			// will only increase as we release more versions.
+			//
+			// This logic has tried to handle multiple
+			// responsibilities. IMO this particular informer
+			// handler should act only as a **switch** to continue
+			// to handle this change further or just reject this
+			// change.
+			//
+			// We need to do a good job to categorise these
+			// **if** conditions. Only the reject related conditions
+			// should be here. Other conditions should be part
+			// of actual business logic that handles change to
+			// a resource.
+			UpdateFunc: func(old, new interface{}) {
+				newCVR := new.(*apis.CStorVolumeReplica)
+				oldCVR := old.(*apis.CStorVolumeReplica)
+
+				glog.V(4).Infof(
+					"received informer update event for cvr {%s}",
```

Comment:
can we log this after `IsRightCStorVolumeReplica` check?

Reply:
ok

Reply:
done

Reply:
:+1: